src/simplify_algebra.cpp

```diff
@@ -78,6 +77,162 @@ struct find_mul_conv
+    {
+        return match::name("gather")(
+            match::args(match::name("reshape")(match::arg(0).bind("data")),
+                        match::is_constant().bind("ind")),
+            match::all_of[match::outputs()](match::used_once(), pointwise()));
+    }
+
+    void apply(module& p, match::matcher_result r) const
+    {
+        auto ins     = r.result;
+        auto ins_rsp = r.instructions["data"];
+        auto ins_ind = r.instructions["ind"];
+        auto ins_pw  = ins->outputs().front();
+
+        // resize input shape
+        auto in_shape = ins_rsp->inputs().front()->get_shape();
+
+        // resize output shape
+        auto out_shape = ins->get_shape();
+
+        // check if output shape is multiple of input shape
+        auto in_lens  = in_shape.lens();
+        auto out_lens = out_shape.lens();
+        if(in_lens.size() != out_lens.size())
+        {
+            return;
+        }
+
+        // if ind is not constant, cannot optimize
+        std::vector<int> vec_ind;
+        auto arg_ind = ins_ind->eval();
+        if(arg_ind.empty())
+        {
+            return;
+        }
+        arg_ind.visit([&](auto v) { vec_ind.assign(v.begin(), v.end()); });
+
+        // output shape must be multiple of input shape
+        std::vector<bool> is_multi(in_lens.size());
+        std::transform(
+            in_lens.begin(), in_lens.end(), out_lens.begin(), is_multi.begin(), [](auto x, auto y) {
+                return (y % x == 0);
+            });
+        if(not std::all_of(is_multi.begin(), is_multi.end(), [](auto b) { return b; }))
+        {
+            return;
+        }
+
+        // output must be multiple of inputs
+        std::vector<std::size_t> scales(in_lens.size());
+        std::transform(
+            in_lens.begin(), in_lens.end(), out_lens.begin(), scales.begin(), [](auto x, auto y) {
+                return y / x;
+            });
+        if(std::any_of(scales.begin(), scales.end(), [](auto i) { return i == 0; }))
+        {
+            return;
+        }
+
+        // wrap up shapes for multibroadcast
+        std::vector<int64_t> in_dims(in_lens.begin(), in_lens.end());
+        std::vector<int64_t> out_dims(out_lens.begin(), out_lens.end());
+        for(int ii = static_cast<int>(scales.size()) - 1; ii >= 0; --ii)
+        {
+            if(scales[ii] == 1 or in_dims[ii] == 1)
+            {
+                continue;
+            }
+
+            out_dims[ii] = in_dims[ii];
+            in_dims.insert(in_dims.begin() + ii + 1, 1);
+            out_dims.insert(out_dims.begin() + ii + 1, scales[ii]);
+        }
+
+        auto in_rsp   = ins_rsp->inputs().front();
+        auto rsp_data = p.insert_instruction(
+            ins_rsp, migraphx::make_op("reshape", {{"dims", in_dims}}), in_rsp);
+        auto mb_rsp = p.insert_instruction(
+            ins_rsp, migraphx::make_op("multibroadcast", {{"output_lens", out_dims}}), rsp_data);
+
+        // reshape other inputs of the pointwise operator
```

Comment:
This should be in a seperate matcher.

Reply:
`find_resize` should just transform `gather` to the reshape+broadcast+reshape. Another matcher should find a pointwise binary operator with one input of `reshape` with non-contiguous inputs, and then transform it to reshape the second argument and apply a reshape to the output.

Reply:
separate it into two optimizations.